File: lib/php/httpparser.php
<?
  class RiddlHttpParser {
    private $params;
    private $MULTIPART_CONTENT_TYPES = array(
      #{{{
      'multipart/form-data',
      'multipart/related',
      'multipart/mixed'
      #}}}
    );
    private $FORM_CONTENT_TYPES = array(
      #{{{
      NULL,
      'application/x-www-form-urlencoded'
      #}}}
    );
    private $EOL = "\r\n";
    private $D = '&;';

    function params(){
      return $this->params;
    }  

    private function parse_content($input,$ctype,$content_length,$content_disposition,$content_id,$riddl_type) {
      #{{{
      if ($content_length == 0) return;
      if ($riddl_type == 'simple') $ctype = NULL;
      $mf = preg_match("/ filename=\"?([^\";]*)\"?/i", $content_disposition, $matchesf); # TODO debug
      $mn = preg_match("/ filename=\"?([^\";]*)\"?/i", $content_disposition, $matchesn); # TODO debug
      $filename = $matchesf[1];
      $name = $mn ? $matchesn[1] : $content_id;

      if (!is_null($ctype) || !is_null($filename)) {
        $body = tmpfile(); # TODO debug
      } else {
        $body = '';
      }

      $bufsize = 16384;

      while ($content_length > 0) {
        $c = fread($input,$bufsize < $content_length ? $bufsize : $content_length);
        if (!$c)
          throw new Exception("bad content body");
        $this->write_body($body,$c);
        $content_length -= strlen($c);
      }

      $this->add_to_params($name,$body,$filename,$ctype,NULL);
      #}}}
    }

    private function parse_multipart($input,$content_type,$content_length) {
      #{{{
      preg_match("/\Amultipart\/.*boundary=\"?([^\";,]+)\"?/",$content_type,$matches);
      $boundary = '--' . $matches[1];

      $boundary_size = strlen($boundary) + strlen($this->EOL);
      $content_length -= $boundary_size;
      $status = fread($input,$boundary_size);
      if (!$status == $boundary . $this->EOL)
        throw new Exception("bad content body");

      $rx = "/(?:" . $this->EOL . ")?" . preg_quote($boundary) . "(" . $this->EOL . "|--)/";

      $buf = '';
      $bufsize = 16384;
      while (true) {
        $head = NULL;
        $body = '';
        $filename = NULL; $ctype = NULL; $name = NULL;

        while (!($head && preg_match($rx,$buf))) {
          if (!$head && $i = strpos($buf,$this->EOL . $this->EOL)) {
            $head = substr($buf,0,$i+2); # First \r\n
            $buf = substr($buf,$i+4); # Second \r\n
            
            $mf = preg_match("/Content-Disposition:.* filename=\"?([^\";]*)\"?/i", $head, $matches);
            $filename = $mf ? $matches[1] : NULL;
            $mc = preg_match("/Content-Type: (.*)" . $this->EOL . "/i", $head, $matches);
            $ctype = $mc ? $matches[1] : NULL;
            $md = preg_match("/Content-Disposition:.*\s+name=\"?([^\";]*)\"?/i", $head, $matchesd);
            $mi = preg_match("/Content-ID:\s*([^" . $this->EOL . "]*)/i", $head, $matchesi);

            $name = $md ? $matchesd[1] : ($mi ? $matchesi[1] : 'bullshit');

            if ($ctype || $filename)
              $body = tmpfile();

            continue;
          }

          # Save the read body part.
          if ($head && ($boundary_size+4 < strlen($buf))) {
            $this->write_body($body,substr($buf,0, strlen($buf) - ($boundary_size+4)));
            $buf = substr($buf,$boundary_size+4);
          }

          $c = fread($input,$bufsize < $content_length ? $bufsize : $content_length);
          if (!$c)
            throw new Exception("bad content body");
          $buf .= $c;
          $content_length -= strlen($c);
        }

        # Save the rest.
        preg_match($rx,$buf,$matches);
        if ($i = strpos($buf,$matches[0])) {
          $this->write_body(&$body,substr($buf,0,$i));
          $buf = substr($buf, $i + $boundary_size+2);
          if ($matches[0] == "--")
            $content_length = -1;
        }
        $this->add_to_params($name,$body,$filename,$ctype,$head);

        if (!$buf || $content_length == -1)
          break;
      }
      #}}}
    }

    private function parse_nested_query($qs, $type) {
      #{{{
      if ($qs) {
        $what = preg_split("/[" . $this->D . "] */",$qs);
        foreach ($what as $p) {
          $p = urldecode($p);
          $p = preg_split('/=/',$p,2);
          array_push($this->params,new RiddlParameterSimple($p[0],$p[1],$type));
        }
      }  
      #}}}
    }

    private function write_body($body,$what) {
      #{{{
      if (is_resource($body))
        fwrite($body,$what);
      if (is_string($body))
        $body .= $what;
      #}}}
    }

    private function add_to_params($name,$body,$filename,$ctype,$head) {
      #{{{
      if (!is_null($filename) && $filename == '') {
        # filename is blank which means no file has been selected
      } elseif ($filename && $ctype) {
        # Take the basename of the upload's original filename.
        # This handles the full Windows paths given by Internet Explorer
        # (and perhaps other broken user agents) without affecting
        # those which give the lone filename.
        preg_match("/^(?:.*[:\\\/])?(.*)/m",$filename,$matches);
        $filename = $matches[1];
        array_push($this->params,new RiddlParameterComplex($name,$ctype,$body,$filename,$head));
      } elseif (!$filename && $ctype) {
        # Generic multipart cases, not coming from a form
        array_push($this->params,new RiddlParameterComplex($name,$ctype,$body,NULL,$head));
      } else {
        array_push($this->params,new RiddlParameterSimple($name,$body,'body'));
      }
      #}}}
    }

    function __construct($query_string,$input,$content_type,$content_length,$content_disposition,$content_id,$riddl_type) {
      #{{{
      $this->params = array();

      $ct = preg_split("/\s*[;,]\s*/",$content_type,2);
      $media_type = strtolower($ct[0]);
      $params = array();
      $this->parse_nested_query($query_string,'query');
      if (array_search($media_type,$this->MULTIPART_CONTENT_TYPES)) {
        $this->parse_multipart($input,$content_type,intval($content_length));
      } elseif (array_search($media_type,$this->FORM_CONTENT_TYPES)) {
        # sub is a fix for Safari Ajax postings that always append \0
        $contents = '';
        while (!feof($input)) {
          $contents .= fread($input, 8192);
        }
        $this->parse_nested_query(preg_replace("/\\0\\z/", '', $contents),'body');
      } else {
        $this->parse_content($input,$content_type,intval($content_length),$content_disposition ? $content_disposition : '',$content_id ? $content_id : '',$riddl_type ? $riddl_type : '');
      }
      #}}}
    }
  }
?>

File: lib/ruby/httpparser.rb
require ::File.dirname(__FILE__) + "/parameter"

module Riddl
  class HttpParser
    MULTIPART_CONTENT_TYPES = [
      #{{{
      'multipart/form-data',
      'multipart/related',
      'multipart/mixed'
      #}}}
    ]
    FORM_CONTENT_TYPES = [
      #{{{
      nil,
      'application/x-www-form-urlencoded'
      #}}}
    ]  
    EOL = "\r\n"
    D = '&;'

    def unescape(s)
      #{{{
      s.tr('+', ' ').gsub(/((?:%[0-9a-fA-F]{2})+)/n){
        [$1.delete('%')].pack('H*')
      }
      #}}}
    end
    private :unescape

    def parse_content(input,ctype,content_length,content_disposition,content_id)
      #{{{
      ctype = nil if ctype == 'text/riddl-data'
      filename = content_disposition[/ filename="?([^\";]*)"?/ni, 1]
      name = content_disposition[/ name="?([^\";]*)"?/ni, 1] || content_id

      if ctype || filename
        body = Parameter::Tempfile.new("RiddlMultipart")
        body.binmode if body.respond_to?(:binmode)
      else
        body = ''
      end
      
      bufsize = 16384
         
      until content_length <= 0
        c = input.read(bufsize < content_length ? bufsize : content_length)
        raise EOFError, "bad content body"  if c.nil? || c.empty?
        body << c
        content_length -= c.size
      end  

      add_to_params(name,body,filename,ctype,nil)
      #}}}
    end
    private :parse_content

    def parse_multipart(input,content_type,content_length)
      #{{{
      content_type =~ %r|\Amultipart/.*boundary=\"?([^\";,]+)\"?|n
      boundary = "--#{$1}"

      boundary_size = boundary.size + EOL.size
      content_length -= boundary_size
      status = input.read(boundary_size)
      raise EOFError, "bad content body" unless status == boundary + EOL

      rx = /(?:#{EOL})?#{Regexp.quote boundary}(#{EOL}|--)/n

      buf = ""
      bufsize = 16384
      loop do
        head = nil
        body = ''
        filename = ctype = name = nil

        until head && buf =~ rx
          if !head && i = buf.index(EOL+EOL)
            head = buf.slice!(0, i+2) # First \r\n
            buf.slice!(0, 2)          # Second \r\n

            filename = head[/Content-Disposition:.* filename="?([^\";]*)"?/ni, 1]
            ctype = head[/Content-Type: (.*)#{EOL}/ni, 1]
            name = head[/Content-Disposition:.*\s+name="?([^\";]*)"?/ni, 1] || head[/Content-ID:\s*([^#{EOL}]*)/ni, 1]

            if ctype || filename
              body = Parameter::Tempfile.new("RiddlMultipart")
              body.binmode  if body.respond_to?(:binmode)
            end

            next
          end

          # Save the read body part.
          if head && (boundary_size+4 < buf.size)
            body << buf.slice!(0, buf.size - (boundary_size+4))
          end

          c = input.read(bufsize < content_length ? bufsize : content_length)
          raise EOFError, "bad content body"  if c.nil? || c.empty?
          content_length -= c.size
          buf << c
        end

        # Save the rest.
        if i = buf.index(rx)
          body << buf.slice!(0, i)
          buf.slice!(0, boundary_size+2)
          content_length = -1  if $1 == "--"
        end

        add_to_params(name,body,filename,ctype,head)

        break if buf.empty? || content_length == -1
      end
      #}}}
    end
    private :parse_multipart

    def add_to_params(name,body,filename,ctype,head)
      #{{{
      if filename == ""
        # filename is blank which means no file has been selected
      elsif filename && ctype
        body.rewind

        # Take the basename of the upload's original filename.
        # This handles the full Windows paths given by Internet Explorer
        # (and perhaps other broken user agents) without affecting
        # those which give the lone filename.
        filename =~ /^(?:.*[:\\\/])?(.*)/m
        filename = $1

        @params << Parameter::Complex.new(name, ctype, body, filename, head)
      elsif !filename && ctype
        body.rewind
        
        # Generic multipart cases, not coming from a form
        @params << Parameter::Complex.new(name, ctype, body, nil, head)
      else
        @params << Parameter::Simple.new(name, body, :body)
      end
      #}}}
    end
    private :add_to_params

    def parse_nested_query(qs, type)
      #{{{
      (qs || '').split(/[#{D}] */n).each do |p|
        k, v = unescape(p).split('=', 2)
        @params << Parameter::Simple.new(k,v,type)
      end
      #}}}
    end
    private :parse_nested_query

    def initialize(query_string,input,content_type,content_length,content_disposition,content_id)
      #{{{
      # rewind because in some cases it is not at start (when multipart without length)
      begin
        input.rewind if input.respond_to?(:rewind)
      rescue Errno::ESPIPE
        # Handles exceptions raised by input streams that cannot be rewound
        # such as when using plain CGI under Apache
      end

      media_type = content_type && content_type.split(/\s*[;,]\s*/, 2).first.downcase
      @params = []
      parse_nested_query(query_string,:query)
      if MULTIPART_CONTENT_TYPES.include?(media_type)
        parse_multipart(input,content_type,content_length.to_i)
      elsif FORM_CONTENT_TYPES.include?(media_type)
        # sub is a fix for Safari Ajax postings that always append \0
        parse_nested_query(input.read.sub(/\0\z/, ''),:body)
      else 
        parse_content(input,content_type,content_length.to_i,content_disposition||'',content_id||'')
      end

      begin
        input.rewind if input.respond_to?(:rewind)
      rescue Errno::ESPIPE
        # Handles exceptions raised by input streams that cannot be rewound
        # such as when using plain CGI under Apache
      end
      #}}}
    end

    attr_reader :params
  end
end
